### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by mkalcok**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,11 +2,33 @@
...
+    $ juju deploy charmed-kubernetes
+    $ juju deploy -n 3 ceph-mon
+    $ juju deploy -n 3 ceph-osd
+    $ juju add-relation ceph-osd ceph-mon
+    
+Once the deployment settled, We can add `ceph-csi`
+
+    $ juju deploy ./ceph-csi.charm
+    $ juju add-relation kubernetes-master
+    $ juju add-relation ceph-mon
```
Comment (auria) [inline README.md]: These 2 commands seem to be missing the `ceph-csi` application for the relations to happen.

File: metadata.yaml [hunk 2]
```diff
@@ -2,18 +2,17 @@
...
-# TEMPLATE-TODO: replace with containers for your workload (delete for non-k8s)
-containers:
-  httpbin:
-    resource: httpbin-image
-
-# TEMPLATE-TODO: each container defined above must specify an oci-image resource
-resources:
-  httpbin-image:
-    type: oci-image
-    description: OCI image for httpbin (kennethreitz/httpbin)
```
Comment (auria) [inline metadata.yaml:18]: I would suggest adding the Ubuntu series this charm in supported in (probably, the same series as kubernetes-master).

File: src/charm.py [hunk 3]
```diff
@@ -12,92 +12,292 @@
...
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
+    @staticmethod
+    def copy_stored_dict(stored_dict: StoredDict) -> dict:
+        """Return copy of StoredDict as a basic dict.
 
-        Learn more about Pebble layers at https://github.com/canonical/pebble
+        This is a convenience function as the StoredDict object does not implement
+        copy() method.
```
Comment (auria) [inline src/charm.py:121]: Should a bug be reported for StoredDict to support a copy() method?

File: src/charm.py [hunk 4]
```diff
@@ -12,92 +12,292 @@
...
+
+        Example:
+            for resource in self.resources:
+                try:
+                    resource.remove()
+                except ApiException as exc:
+                    if exc.status == 404:
+                        pass  # Resource already deleted.
+                    else:
+                        raise exc
```
Comment (auria) [inline src/charm.py:152]: The if-else could be simplified by:
```
# If 404, resource already deleted
if exc.status != 404:
    raise exc
```

File: src/charm.py [hunk 5]
```diff
@@ -12,92 +12,292 @@
...
+            func(self, *args, **kwargs)
+
+    return leader_check
+
+
 class CephCsiCharm(CharmBase):
     """Charm the service."""
 
+    CEPH_RELATION = "ceph"
+    K8S_NS = "default"
```
Comment (auria) [inline src/charm.py:81]: I imagine this is the behavior in the cdk-addon snap. It would be worth pointing it out somewhere, in case the k8s team may consider making this value custom (e.g. via `config.yaml`)

File: src/charm.py [hunk 6]
```diff
@@ -12,92 +12,292 @@
...
-        and the actions.py file.
+    def update_stored_state(self, config_name: str, stored_state_name: str) -> bool:
+        """Update value in stored stated based on the value from config.
 
-        Learn more about actions at https://juju.is/docs/sdk/actions
+        Since ConfigChangedEvent does not provide information about which config options
+        were updated, this convenience method allows you to pass name of the config
+        option and name of the corresponding variable in StoredState. If the values
+        do not match, StoredState variable is updated with new value and `True` is
+        returned.
```
Comment (auria) [inline src/charm.py:190]: Do you think this should be reported for the Operator Framework maintainers consideration?

File: src/charm.py [hunk 7]
```diff
@@ -12,92 +12,292 @@
...
+        """Execute "on install" event callback."""
+        self.check_required_relations()
+
+    def check_required_relations(self) -> None:
+        """Run check if any required relations are missing"""
+        if self.model.get_relation(self.CEPH_RELATION) is None:
+            self.unit.status = BlockedStatus("Missing relations: ceph")
         else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+            self.unit.status = ActiveStatus()
```
Comment (auria) [inline src/charm.py]: A convention in OpenStack charms is to use `ActiveStatus("Unit is ready")`, since the status would be empty [1] in this case. If the unit is not ready, because e.g. k8s resources need to be created, then `WaitingStatus` or `MaintenanceStatus` should be used.

1. https://github.com/canonical/operator/blob/b1a8fffc053fbbdaf43b826e78c220bfdae4c8fe/ops/model.py#L890

File: src/charm.py [hunk 8]
```diff
@@ -12,92 +12,292 @@
...
+
+    @needs_leader
+    def create_ceph_resources(self, resources: List[Dict]) -> None:  # pylint: disable=R0201
```
Comment (auria) [inline src/charm.py:227]: Do you think a library with helper functions should be created for functions like this one? I checked the pylint error, and that's why I suggest that approach: using helper functions out of `src/charm.py` when possible.

I also wonder if an external lib (similar to the reactive framework layers) could be used. Some code that could be imported (via `charmcraft create-lib`, `charmcraft fetch-lib`, etc.) and that you would see as reusable in multiple charms.

File: src/charm.py [hunk 9]
```diff
@@ -12,92 +12,292 @@
...
+        """Render resource definitions from templates in self.RESOURCE_TEMPLATES."""
+        env = Environment(loader=FileSystemLoader(self.template_dir))
+
+        resources = [
+            env.get_template(template).render(self._stored.ceph_data)
+            for template in self.RESOURCE_TEMPLATES
+        ]
+
+        resource_dicts = [yaml.safe_load_all(res) for res in resources]
+        return list(itertools.chain.from_iterable(resource_dicts))
```
Comment (auria) [inline src/charm.py:253]: I wonder about the `resources` variable and doing 2 for-loops rather than one. I agree it is readable as it is, but would it be less readable in a single for-loop?

File: src/charm.py [hunk 10]
```diff
@@ -12,92 +12,292 @@
...
+        resource = env.get_template(self.STORAGE_CLASS_TEMPLATE).render(ext4_ctx)
+        storage_classes.append(yaml.safe_load(resource))
+
+        xfs_ctx = self.copy_stored_dict(self._stored.ceph_data)
+        xfs_ctx["default"] = default_storage == self.XFS_STORAGE
+        xfs_ctx["pool_name"] = "xfs-pool"
+        xfs_ctx["fs_type"] = "xfs"
+        xfs_ctx["sc_name"] = "ceph-xfs"
+        resource = env.get_template(self.STORAGE_CLASS_TEMPLATE).render(xfs_ctx)
+        storage_classes.append(yaml.safe_load(resource))
```
Comment (auria) [inline src/charm.py:275]: Should this be refactored to avoid repeating code? Maybe a function that returns the rendered resource?

```
# new class attribute
SUPPORTED_STORAGE = [EXT4_STORAGE, XFS_STORAGE]
# ...
for storage_type in self.SUPPORTED_STORAGE:
    # remove the "ceph-" part from "ceph-<format>"
    storage_format = storage_type.split('-', 1)[1]
    context = self.copy_stored_dict(...)
    context.update({
        "default": default_storage == storage_type,
        "pool_name": "{}-pool".format(storage_format),
        ....
    })
    resource = self._render_resource(self.STORAGE_CLASS_TEMPLATE, context)
   # etc.
```

File: src/charm.py [hunk 11]
```diff
@@ -12,92 +12,292 @@
...
+        return self.render_resource_definitions() + self.render_storage_definitions()
+
+    @needs_leader
+    def update_storage_classes(self) -> None:
+        """Re-render templates and update StorageClass resources in k8s cluster."""
+        for resource in self.resources:
+            if isinstance(resource, StorageClass):
+                resource.remove()
+        storage_classes = self.render_storage_definitions()
+        self.create_ceph_resources(storage_classes)
```
Comment (auria) [inline src/charm.py]: I wonder about adding 3 lines of inline comments to separate helper functions and hooks-related functions. This is completely nitpick, since everything ends in the same class definition.

File: src/charm.py [hunk 12]
```diff
@@ -12,92 +12,292 @@
...
+
+    @needs_leader
+    def _on_ceph_joined(self, event: RelationJoinedEvent) -> None:
+        """Create necessary k8s resources when relation is formed with ceph-mon."""
+        if self._stored.resources_created:
+            # Skip silently if other ceph_relation_joined event already
+            # created resources
+            return
```
Comment (auria) [inline src/charm.py:355]: Could this be logged? It would help operations troubleshoot if relations were not created (or maybe accidentally removed manually, or out of Juju's control).

File: src/charm.py [hunk 13]
```diff
@@ -12,92 +12,292 @@
...
+        unit_data = event.relation.data[event.unit]
+        expected_data = ("fsid", "key", "mon_hosts")
+        for key in expected_data:
+            self._stored.ceph_data[key] = unit_data.get(key)
+        missing_data = [key for key, value in self._stored.ceph_data.items() if value is None]
+        if missing_data:
+            logger.warning(
+                "Ceph relation with %s is missing data: %s", event.unit.name, missing_data
+            )
+            self.unit.status = BlockedStatus("Ceph relation is missing data.")
```
Comment (auria) [inline src/charm.py]: Should this be `WaitingStatus`? If a user needs to run any step against this unit, then `blocked` is accurate. If the user needs to wait for other processes within Juju to be completed (e.g. queue hooks that need to be triggered in this unit), then `waiting` should be the status.

Could this happen because the Ceph units are being bootstrapped / configured, so they're not ready to serve other applications?

File: src/charm.py [hunk 14]
```diff
@@ -12,92 +12,292 @@
...
+                    "Removing resource %s (namespace=%s)",
+                    resource.name,
+                    (resource.namespace or None),
+                )
+                resource.remove()
+            except ApiException as exc:
+                if exc.status == 404:
+                    logger.info("Resource %s is already removed.", resource.name)
+                else:
+                    raise exc
```
Comment (auria) [inline src/charm.py:428]: Same nitpick comment as previously.

File: src/resource.py [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+from kubernetes.client import AppsV1Api, CoreV1Api
+from kubernetes.client import RbacAuthorizationV1Api as RbacAuthApi
+from kubernetes.client import StorageV1Api
+
+
+class MissingMethod(BaseException):
+    """Exception that represents method that is not implemented."""
+
+
+class Resource:
```
Comment (auria) [inline src/resource.py:28]: Should this be called `KubernetesResourceBase`?

File: src/resource.py [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+
+    def remove(self) -> None:
+        """Call appropriate api method to remove k8s resource."""
+        if self.namespace:
+            self._remove_namespaced_action(self.name, self.namespace)
+        else:
+            self._remove_action(self.name)
+
+
+class CoreResource(Resource):
```
Comment (auria) [inline src/resource.py:131]: Similar to the previous comment, should this class be named `KubernetesCoreResource` ?

File: src/resource.py [hunk 17]
```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
```
Comment (auria) [inline src/resource.py:1]: I wonder if this file should live in `lib/` instead of `src/`, and if the file should be renamed to `kubernetes_resources.py`

File: src/charm.py [hunk 18]
```diff
@@ -12,92 +12,292 @@
...
 logger = logging.getLogger(__name__)
 
 
+def needs_leader(func: Callable) -> Callable:
+    """Ensure that function with this decorator is executed only if on leader units."""
+
+    @wraps(func)
+    def leader_check(self: CharmBase, *args: Any, **kwargs: Any) -> Any:
+        if self.unit.is_leader():
+            func(self, *args, **kwargs)
```
Comment (auria) [inline src/charm.py]: Could this function log when `Unit is not the leader. Skipping execution of {func}`

**Review by auria — CHANGES_REQUESTED** [review]
Overall, this new charm looks great.

I have added several comments inline. The most important ones are about logging and using helpers in libs rather than within `src/charm.py` (or `src/`).

What would happen if a user relates this subordinate charm to 2 different ceph-mon applications? (like, 2 Ceph clusters). If this charm only supports relations to a single ceph-mon application, it would be worth exploring how to make it clear (in docs, but also via limitations in metadata.yaml or evaluating relation-ids, if len(relids) > 1).

I also understand that unit/func tests would be developed in the future, and I would recommend using GH workflows. The main doubt I have is if ceph-relation-broken will be able to delete all the k8s resources or if there could be corner cases that would leave resources behind. This is worth asking/exploring with the k8s team (and even the Juju team) for their recommendation.

File: src/charm.py (hunk 3 above)
Comment (mkalcok) [reply src/charm.py:121]: I opened this issue in Operator framework, https://github.com/canonical/operator/issues/587 we can see what the maintainers will say.

File: src/charm.py (hunk 4 above)
Comment (mkalcok) [reply src/charm.py:152]: Duh! of course...thanks :D

File: src/charm.py (hunk 5 above)
Comment (mkalcok) [reply src/charm.py:81]: Opened #2

File: src/charm.py (hunk 6 above)
Comment (mkalcok) [reply src/charm.py:190]: This sounds either like conscious decision or that there's some other limitation. Maybe because the `ConfigChangedEvent` does not necessarily have to be triggered by direct call to `juju config`([see config-changed section](https://juju.is/docs/sdk/events)). But maybe I can open a conversation with the Ops team about this.

File: src/charm.py (hunk 7 above)
Comment (mkalcok) [reply src/charm.py]: Status message makes sense :+1: 

Regarding the `WaitingStatus`, from what I saw, the resource creation is pretty quick, no waiting time. Unless we want to add more complex check method that would verify status of all the expected resources (not saying it's a bad thing), I don't see a use case for `WaitingStatus`

File: src/charm.py (hunk 8 above)
Comment (mkalcok) [reply src/charm.py:227]: this one is bit tricky, but the `pylint` error is sort of false-positive. `needs_leader` decorator can really only wrap methods of `CharmBase` class instances. It expects first argument (self) to be reference to `CharmBase` instance and then it internally uses `self.is_leader()`

File: src/charm.py (hunk 9 above)
Comment (mkalcok) [reply src/charm.py:253]: Do you mean replacing these two lines with nested list comprehension? Because I always find them extra confusing :D

File: src/charm.py (hunk 10 above)
Comment (mkalcok) [reply src/charm.py:275]: This method will likely see refactor when `cephfs` support will be added.

File: src/charm.py (hunk 12 above)
Comment (mkalcok) [reply src/charm.py:355]: I opened #3 about adding thorough logging to the whole charm.

File: src/charm.py (hunk 13 above)
Comment (mkalcok) [reply src/charm.py]: Good point, this should be `WaitingStatus`. I also opened #4 because we need a way to recover from this `Waiting` state.

File: src/charm.py (hunk 11 above)
Comment (mkalcok) [reply src/charm.py]: Sorry, I don't understand this one.

Comment (auria) [reply src/charm.py]: This is more of a visual OCD. The suggestion goes along the lines of adding:

```
#
# Helper functions
#
def ...

#
# Observed functions (triggered by hooks)
#
def _on_config_changed(...):
```

Ideally, the helper functions could be part of a lib which accepts arguments with class attributes (so the calls to self.helper_method wouldn't be needed).

Again, please ignore if you don't find this helpful :)

File: src/charm.py (hunk 18 above)
Comment (mkalcok) [reply src/charm.py]: I was thinking about this, but there's not really a good way to get "human readable" name of the function. We could use `func.__name__` but that really exposes internal implementation and might not be easy to understand without knowing context (code).
We can certainly do it, but I thought that it does add that much value to have non-leader units announce that the are NOT doing something :D

File: src/resource.py (hunk 15 above)
Comment (mkalcok) [reply src/resource.py:28]: It's a mouthful, but I don't mind. Do you also want to change naming for specific resources, like `Secret` -> `KubernetesSecret` ? Because I think that might be too much. If you'll have something like
```python
from kubernetes_resources import Secret
```
I think it gives enough context.

File: src/charm.py (hunk 18 above)
Comment (auria) [reply src/charm.py]: leadership may change for different reasons. A unit may show as leader but have logs about not being the leader (from a few minutes or hours ago).

OTOH, support and operations teams use juju-crashdump which collect all the Juju logs from a Juju model. In debug mode, Juju will tell if a unit is the leader or not. But I think it would not hurt to make the unit itself log about it, which from my experience is more useful than the Juju modules logs.

File: src/resource.py (hunk 15 above)
Comment (auria) [reply src/resource.py:28]: you're raising an interesting point. I do agree that `Secret` could be left as is. A charm developer could use `import Secret as KubernetesSecret` if another `Secret` definition could exist in the future.
